application/ui/src/features/annotator/tools/segment-anything-tool/segment-anything-tool.component.tsx

```diff
@@ -1,87 +1,111 @@
+                    strokeWidth={'calc(3px / var(--zoom-scale))'}
+                    fill={'transparent'}
+                    className={classes.animateStroke}
+                >
+                    <AnnotationShape annotation={annotation} />
+                </g>
+            ))}
+        </MaskAnnotations>
+    );
 };
 
 export const SegmentAnythingTool = () => {
-    const [mousePosition, setMousePosition] = useState<InteractiveAnnotationPoint>();
+    const [_createLabelFormPosition, setCreateLabelFormPosition] = useState<Point | null>(null);
     const [previewShapes, setPreviewShapes] = useState<Shape[]>([]);
+    const [acceptedShapes, setAcceptedShapes] = useState<Shape[] | null>(null);
+    const ref = useRef<SVGSVGElement>(null);
 
     const zoom = useZoom();
-    const { mediaItem, roi, image } = useAnnotator();
+    const { roi, image, selectedLabel } = useAnnotator();
+    const { addAnnotations } = useAnnotationActions();
     const { isLoading, decodingQueryFn } = useSegmentAnythingModel();
     const throttledDecodingQueryFn = useSingleStackFn(decodingQueryFn);
-    const decodingMutation = useDecodingMutation(decodingQueryFn);
 
-    const ref = useRef<SVGRectElement>(null);
+    const canvasRef = useRef<SVGRectElement>(null);
 
     const clampPoint = clampPointBetweenImage(image);
 
-    const throttleSetMousePosition = useThrottledCallback((point: InteractiveAnnotationPoint) => {
-        setMousePosition(point);
-    }, THROTTLE_TIME);
+    const handleMouseMove = (event: PointerEvent<SVGSVGElement>) => {
+        if (acceptedShapes !== null) {
+            return;
+        }
 
-    useEffect(() => {
-        if (mousePosition === undefined) {
+        if (!canvasRef.current) {
             return;
         }
 
-        throttledDecodingQueryFn([mousePosition])
+        const point = clampPoint(
+            getRelativePoint(canvasRef.current, { x: event.clientX, y: event.clientY }, zoom.scale)
+        );
+
+        throttledDecodingQueryFn([{ ...point, positive: true }])
             .then((shapes) => {
                 setPreviewShapes(shapes.map((shape) => removeOffLimitPoints(shape, roi)));
-
-                throttleSetMousePosition.flush();
             })
             .catch(() => {
                 // If getting decoding went wrong we set an empty preview and
                 // start to compute the next decoding
                 return [];
             });
-    }, [mousePosition, throttledDecodingQueryFn, throttleSetMousePosition, roi]);
+    };
 
-    const handleMouseMove = (event: PointerEvent<SVGSVGElement>) => {
-        if (!ref.current) {
-            return;
-        }
+    const handleAddAnnotations = (shapes: Shape[], label: Label) => {
+        addAnnotations(shapes, [label]);
+        setPreviewShapes([]);
+    };
 
-        const point = clampPoint(getRelativePoint(ref.current, { x: event.clientX, y: event.clientY }, zoom.scale));
+    // const handleAddAnnotationsCreateLabel = (label: Label) => {
```

Comment:
I think for Tune we should not make SAM be responsible for creating new labels. I imagine that for Tune it will be more expensive to add labels (users would need to retrain their models before inference will return these labels) so having a separate more dedicated UI would feel more appropriate to me.

If we do want an easy way to add new labels from the annotator screen then I'd prefer us to think of a design that would allow us to reuse the same design for the other tools (bounding box, polygon etc) as otherwise that could get inconsistent.